vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1349 @@
+		"sub": tok.Subject,
+		"aud": tok.Audience,
+		"exp": tok.Expiry,
+		"iat": tok.IssuedAt,
+	}
+
+	// Parse and integrate the populated role template. Structural errors with the template _should_
+	// be caught during role configuration. Error found during runtime will be logged, but they will
+	// not block generation of the basic ID token. They should not be returned to the requester.
+	_, populatedTemplate, err := identity.PopulateString(&identity.PopulateStringInput{
+		Mode:   identity.JSONTemplating,
+		String: template,
+		Entity: entity,
+		Groups: groups,
+		// namespace?
+	})
+
+	if err != nil {
+		logger.Warn("error populating OIDC token template", "template", template, "error", err)
+	}
+
+	if populatedTemplate != "" {
+		var parsed map[string]interface{}
+		if err := json.Unmarshal([]byte(populatedTemplate), &parsed); err != nil {
+			logger.Warn("error parsing OIDC template", "template", template, "err", err)
+		}
+
+		for k, v := range parsed {
+			if !strutil.StrListContains(requiredClaims, k) {
+				output[k] = v
+			} else {
+				logger.Warn("invalid top level OIDC template key", "template", template, "key", k)
+			}
+		}
+	}
+
+	payload, err := json.Marshal(output)
+	if err != nil {
+		return nil, err
+	}
+
+	return payload, nil
+}
+
+func (k *namedKey) signPayload(payload []byte) (string, error) {
+	signingKey := jose.SigningKey{Key: k.SigningKey, Algorithm: jose.SignatureAlgorithm(k.Algorithm)}
+	signer, err := jose.NewSigner(signingKey, &jose.SignerOptions{})
+	if err != nil {
+		return "", err
+	}
+
+	signature, err := signer.Sign(payload)
+	if err != nil {
+		return "", err
+	}
+
+	signedIdToken, err := signature.CompactSerialize()
+	if err != nil {
+		return "", err
+	}
+
+	return signedIdToken, nil
+}
+
+func (i *IdentityStore) pathOIDCRoleExistenceCheck(ctx context.Context, req *logical.Request, d *framework.FieldData) (bool, error) {
+	role, err := i.getOIDCRole(ctx, req.Storage, d.Get("name").(string))
+	if err != nil {
+		return false, err
+	}
+
+	return role != nil, nil
+}
+
+// handleOIDCCreateRole is used to create a new role or update an existing one
+func (i *IdentityStore) pathOIDCCreateUpdateRole(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	var role role
+	entry, err := req.Storage.Get(ctx, roleConfigPath+name)
+	if err != nil {
+		return nil, err
+	}
+	if entry != nil {
+		if err := entry.DecodeJSON(&role); err != nil {
+			return nil, err
+		}
+	}
+
+	if key, ok := d.GetOk("key"); ok {
+		role.Key = key.(string)
+	} else if req.Operation == logical.CreateOperation {
+		role.Key = d.Get("key").(string)
+	}
+
+	if role.Key == "" {
+		return logical.ErrorResponse("key must be provided"), nil
+	}
+
+	// validate that key exists
+	entry, err = req.Storage.Get(ctx, namedKeyConfigPath+role.Key)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return logical.ErrorResponse("Key %q does not exist", role.Key), nil
```

Comment:
Here and below, please un-capitalize the error

Reply:
Fixed